pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/decoder/protobufnative/PulsarProtobufNativeColumnDecoder.java

```diff
@@ -0,0 +1,386 @@
+import io.prestosql.spi.type.RowType;
+import io.prestosql.spi.type.RowType.Field;
+import io.prestosql.spi.type.SmallintType;
+import io.prestosql.spi.type.TimestampType;
+import io.prestosql.spi.type.TinyintType;
+import io.prestosql.spi.type.Type;
+import io.prestosql.spi.type.VarbinaryType;
+import io.prestosql.spi.type.VarcharType;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Pulsar {@link org.apache.pulsar.common.schema.SchemaType#PROTOBUF_NATIVE} ColumnDecoder.
+ */
+public class PulsarProtobufNativeColumnDecoder {
+    private static final Set<Type> SUPPORTED_PRIMITIVE_TYPES = ImmutableSet.of(
+            BooleanType.BOOLEAN,
+            IntegerType.INTEGER,
+            BigintType.BIGINT,
+            RealType.REAL,
+            DoubleType.DOUBLE,
+            VarbinaryType.VARBINARY);
+
+    private final Type columnType;
+    private final String columnMapping;
+    private final String columnName;
+
+    public PulsarProtobufNativeColumnDecoder(DecoderColumnHandle columnHandle) {
+        try {
+            requireNonNull(columnHandle, "columnHandle is null");
+            this.columnType = columnHandle.getType();
+            this.columnMapping = columnHandle.getMapping();
+            this.columnName = columnHandle.getName();
+            checkArgument(!columnHandle.isInternal(),
+                    "unexpected internal column '%s'", columnName);
+            checkArgument(columnHandle.getFormatHint() == null,
+                    "unexpected format hint '%s' defined for column '%s'", columnHandle.getFormatHint(), columnName);
+            checkArgument(columnHandle.getDataFormat() == null,
+                    "unexpected data format '%s' defined for column '%s'", columnHandle.getDataFormat(), columnName);
+            checkArgument(columnHandle.getMapping() != null,
+                    "mapping not defined for column '%s'", columnName);
+            checkArgument(isSupportedType(columnType),
+                    "Unsupported column type '%s' for column '%s'", columnType, columnName);
+        } catch (IllegalArgumentException e) {
+            throw new PrestoException(GENERIC_USER_ERROR, e);
+        }
+    }
+
+    private boolean isSupportedType(Type type) {
+        if (isSupportedPrimitive(type)) {
+            return true;
+        }
+
+        if (type instanceof ArrayType) {
+            checkArgument(type.getTypeParameters().size() == 1,
+                    "expecting exactly one type parameter for array");
+            return isSupportedType(type.getTypeParameters().get(0));
+        }
+
+        if (type instanceof MapType) {
+            List<Type> typeParameters = type.getTypeParameters();
+            checkArgument(typeParameters.size() == 2,
+                    "expecting exactly two type parameters for map");
+            return isSupportedType(type.getTypeParameters().get(1)) && isSupportedType(type.getTypeParameters().get(0));
```

Comment:
A trivial point, maybe we could use the the `typeParameters.get(1) instead of the `type.getTypeParameters().get(1)`.